tslearn/hdftools.py

```diff
@@ -0,0 +1,152 @@
+import h5py
+import os
+import numpy as np
+import traceback
+from warnings import warn
+
+
+def save_dict(d, filename, group, raise_type_fail=True):
+    """
+    Recursively save a dict to an hdf5 group in a new file.
+
+    Parameters
+    ----------
+    d: dict
+        dict to save as an hdf5 file
+
+    filename : str
+        Full path to save the file to. File must not already exist.
+
+    group : str
+        group name to save the dict to
+
+    raise_type_fail : bool
+        If True: raise an exception if saving a part of the dict fails.
+        If False: prints a warning instead and saves the
+        object's __str__() return value.
+
+    Returns
+    -------
+    None
+
+    Raises
+    ------
+    FileExistsError
+        If the path specified by the `filename` parameter already exists.
+
+    TypeError
+        If a particular entry within the dict cannot be saved to hdf5 AND
+        the argument `raise_type_fail` is set to `True`
+    """
+
+    if os.path.isfile(filename):
+        raise FileExistsError
+
+    with h5py.File(filename, 'w') as h5file:
+        _dicts_to_group(h5file, "{}/".format(group), d,
+                        raise_meta_fail=raise_type_fail)
+
+
+def _dicts_to_group(h5file, path, d, raise_meta_fail):
+
+    for key, item in d.items():
+
+        if isinstance(item, np.ndarray):
+
+            if item.dtype == np.dtype('O'):
+                # see if h5py is ok with it
+                try:
+                    h5file[path + key] = item
+                    # h5file[path + key].attrs['dtype'] = item.dtype.str
+                except:
+                    msg = "numpy dtype 'O' for item: {}" \
+                          "not supported by HDF5\n{}" \
+                          "".format(item, traceback.format_exc())
+
+                    if raise_meta_fail:
+                        raise TypeError(msg)
+                    else:
+                        h5file[path + key] = str(item)
+                        warn("{}, storing whatever str(obj) returns"
+                             "".format(msg))
+
+            # numpy array of unicode strings
+            elif item.dtype.str.startswith('<U'):
+                h5file[path + key] = item.astype(h5py.special_dtype(vlen=str))
+
+                # otherwise h5py doesn't restore the right dtype for str types
+                h5file[path + key].attrs['dtype'] = item.dtype.str
+
+            # other types
+            else:
+                h5file[path + key] = item
+                # h5file[path + key].attrs['dtype'] = item.dtype.str
+
+        # single pieces of data
+        elif isinstance(item, (str, bytes, int, float, np.int, np.int8,
```

Comment:
You can use parent classes if you don't want to add them all manually (e.g. replacing np.int, ..., np.int64 with np.integer), but there are some differences (for instance unsigned integers):
https://docs.scipy.org/doc/numpy/reference/arrays.scalars.html